src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -102,159 +128,60 @@ public abstract class TextEncoder
-            out int charsConsumed,
-            out int charsWritten,
-            int firstCharacterToEncode,
-            bool isFinalBlock = true)
+        private string EncodeToNewString(ReadOnlySpan<char> value, int indexOfFirstCharToEncode)
         {
-            Debug.Assert(value != null);
-            Debug.Assert(firstCharacterToEncode >= 0);
-
-            char* originalBuffer = buffer;
-            charsWritten = 0;
-
-            if (firstCharacterToEncode > 0)
-            {
-                Debug.Assert(firstCharacterToEncode <= valueLength);
-                Buffer.MemoryCopy(source: value,
-                    destination: buffer,
-                    destinationSizeInBytes: sizeof(char) * bufferLength,
-                    sourceBytesToCopy: sizeof(char) * firstCharacterToEncode);
-
-                charsWritten += firstCharacterToEncode;
-                bufferLength -= firstCharacterToEncode;
-                buffer += firstCharacterToEncode;
-            }
-
-            int valueIndex = firstCharacterToEncode;
-
-            char firstChar = value[valueIndex];
-            char secondChar = firstChar;
-            bool wasSurrogatePair = false;
-
-            // this loop processes character pairs (in case they are surrogates).
-            // there is an if block below to process single last character.
-            int secondCharIndex;
-            for (secondCharIndex = valueIndex + 1; secondCharIndex < valueLength; secondCharIndex++)
-            {
-                if (!wasSurrogatePair)
-                {
-                    firstChar = secondChar;
-                }
-                else
-                {
-                    firstChar = value[secondCharIndex - 1];
-                }
-
-                secondChar = value[secondCharIndex];
-
-                if (!WillEncode(firstChar))
-                {
-                    wasSurrogatePair = false;
-                    *buffer = firstChar;
-                    buffer++;
-                    bufferLength--;
-                    charsWritten++;
-                }
-                else
-                {
-                    int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, secondChar, out wasSurrogatePair, out bool _);
-                    if (!TryEncodeUnicodeScalar(nextScalar, buffer, bufferLength, out int charsWrittenThisTime))
-                    {
-                        charsConsumed = (int)(originalBuffer - buffer);
-                        return OperationStatus.DestinationTooSmall;
-                    }
+            ReadOnlySpan<char> remainingInput = value.Slice(indexOfFirstCharToEncode);
+            ValueStringBuilder stringBuilder = new ValueStringBuilder(stackalloc char[EncodeStartingOutputBufferSize]);
 
-                    if (wasSurrogatePair)
-                    {
-                        secondCharIndex++;
-                    }
+#if !NETCOREAPP
+            // Can't call string.Concat later in the method, so memcpy now.
```

Comment:
I'm not sure I understand, why can't we do this in NETCOREAPP targets?

Reply:
It's "if __not__ netcoreapp", because the `string.Concat(ROS<char>, ...)` overloads don't exist downlevel.

Reply:
I probably didn't phrase the question correctly, why can't we just call the same `ValueStringBuilder.Append(ROS<char>)` method in NETCOREAPP and avoid using `string.Concat`?

Reply:
We can. However, this has possible negative perf impact. It results in an unneeded copy of some chunk of the input string into the temporary storage backing the ValueStringBuilder. It also increases the likelihood that the VSB instance will need to resize its internal buffers at some point. This ifdef allows us to minimize the chances of this happening by ensuring the VSB never deals with irrelevant data.